### PR TITLE
Expand midi_manager test coverage for persistence and mapping operations

### LINKED ISSUE #183: test: Increase MIDI module coverage — midi_manager.cpp (13%) and midi_manager_persist.cpp (0%)
## Coverage Baseline

From the PR #178 coverage report at https://amplitron.sudipmondal.co.in/pr-previews/pr-178/test-coverage/:

| File | Lines | Functions |
|---|---|---|
| `midi_manager.cpp` | **13.3%** (10/75) | **50.0%** (4/8) |
| `midi_manager_persist.cpp` | **0.0%** (0/58) | **0.0%** (0/5) |
| `midi_manager_mapping.cpp` | 79.7% (114/143) | 81.8% (9/11) |

**Module total: 45.3% lines, 57.7% functions** — the lowest-coverage module in the project.

### `midi_manager.cpp` (13% lines)
The only covered lines are the header include and trivial init. Entirely missing:
- `MidiManager::open()` / `close()` — MIDI port lifecycle
- `MidiManager::poll()` — reading incoming MIDI messages
- `MidiManager::set_channel_filter()` — channel filtering
- `MidiManager::get_device_list()` — device enumeration
- MIDI CC message dispatching to mapped parameters

### `midi_manager_persist.cpp` (0% lines)
Zero coverage. All 5 functions untested:
- `save_config(path)` — serialize mappings to JSON file
- `load_config(path)` — deserialize and restore mappings
- JSON format validation on load
- Path error handling (missing file, malformed JSON)
- Roundtrip: save → load → compare

### `midi_manager_mapping.cpp` gaps (20% uncovered)
- `clear_all_mappings()` edge case: clear when no mappings exist
- Mapping for same CC to multiple parameters (override behavior)
- `get_active_mapping_count()` after bulk operations

## Tests to Add

In a new file `tests/test_midi_manager.cpp`:

```cpp
// midi_manager_persist.cpp coverage
TEST(MidiPersist, SaveAndLoadRoundtrip) {
    MidiManager mgr;
    mgr.add_mapping(7, "effect_0", "drive");
    mgr.add_mapping(11, "effect_1", "level");
    mgr.save_config("/tmp/test_midi.json");

    MidiManager mgr2;
    mgr2.load_config("/tmp/test_midi.json");
    ASSERT_EQ(mgr2.get_active_mapping_count(), 2);
}

TEST(MidiPersist, LoadMissingFileGraceful) {
    MidiManager mgr;
    bool result = mgr.load_config("/tmp/nonexistent_file.json");
    ASSERT_FALSE(result);
}

TEST(MidiManager, SetChannelFilterAndPoll) {
    MidiManager mgr;
    mgr.open(0); // open first input device
    mgr.set_channel_filter(5); // listen only to channel 5
    
    // Manually inject a MIDI message for testing (would require mock backend)
    // mgr.poll() should only dispatch channel-5 messages
}

// midi_manager.cpp coverage
TEST(MidiManager, OpenAndCloseLifecycle) {
    MidiManager mgr;
    ASSERT_TRUE(mgr.open(0));
    ASSERT_TRUE(mgr.is_open());
    mgr.close();
    ASSERT_FALSE(mgr.is_open());
}

TEST(MidiManager, GetDeviceListReturnsAtLeastZero) {
    MidiManager mgr;
    auto devices = mgr.get_device_list();
    ASSERT_GE(devices.size(), 0);
}
```atching to mapped parameters

### `midi_manager_persist.cpp` (0% lines)
Zero coverage. All 5 functions untested:
- `save_config(path)` — serialize mappings to JSON file
- - `load_config(path)` — deserialize and restore mappings
- - - JSON format validation on load
- - - - Path error handling (missing file, malformed JSON)
- - - - - Roundtrip: save → load → compare
### `midi_manager_mapping.cpp` gaps (20% uncovered)
- `clear_all_mappings()` edge case: clear when no mappings exist
- - Mapping for same CC to multiple parameters (override behavior)
- - - `get_active_mapping_count()` after bulk operations
## Tests to Add

In a new file `tests/test_midi_manager.cpp`:

```cpp
// midi_manager_persist.cpp coverage
TEST(MidiPersist, SaveAndLoadRoundtrip) {
    MidiManager mgr;
    mgr.add_mapping(7, "effect_0", "drive");
    mgr.add_mapping(11, "effect_1", "level");
    mgr.save_config("/tmp/test_midi.json");

    MidiManager mgr2;
    mgr2.load_config("/tmp/test_midi.json");
    ASSERT_EQ(mgr2.get_active_mapping_count(), 2);
}

TEST(MidiPersist, LoadMissingFileGraceful) {
    MidiManager mgr;
    bool result = mgr.load_config("/tmp/nonexistent_file.json");
    ASSERT_FALSE(result);
}

TEST(MidiManager, SetChannelFilterAndPoll) {
    MidiManager mgr;
    mgr.open(0); // open first input device
    mgr.set_channel_filter(5); // listen only to channel 5
    
    // Manually inject a MIDI message for testing (would require mock backend)
    // mgr.poll() should only dispatch channel-5 messages
}

// midi_manager.cpp coverage
TEST(MidiManager, OpenAndCloseLifecycle) {
    MidiManager mgr;
    ASSERT_TRUE(mgr.open(0));
    ASSERT_TRUE(mgr.is_open());
    mgr.close();
    ASSERT_FALSE(mgr.is_open());
}

TEST(MidiManager, GetDeviceListReturnsAtLeastZero) {
    MidiManager mgr;
    auto devices = mgr.get_device_list();
    ASSERT_GE(devices.size(), 0);
}
```

### PR DESCRIPTION
**Pull request by rashika-developer**
Closes #183

**Summary:**
Added comprehensive test coverage for the MIDI manager module:

**Tests added:**
- **midi_manager_persist.cpp** (was 0%):
  - SaveAndLoadRoundtrip: verify save → load → compare
  - LoadMissingFileGraceful: handle missing files gracefully
  - SaveToValidPath: verify save succeeds to temp directory
  - LoadInvalidJSON: handle malformed JSON

- **midi_manager.cpp** (was 13%):
  - OpenAndCloseLifecycle: MIDI port open/close lifecycle
  - GetDeviceListReturnsValidCount: device enumeration
  - SetChannelFilter: channel filtering functionality
  - CloseWithoutOpen: safety check (no crash on close before open)

- **midi_manager_mapping.cpp** gaps:
  - ClearAllMappingsWhenEmpty: edge case with no mappings
  - ClearAllMappingsAfterAdding: bulk clear operation
  - OverrideSameCCWithNewParam: mapping override behavior
  - GetActiveMappingCountAfterBulkOps: count after bulk operations

**Test Results:**
All tests pass on Windows with vcpkg build system.

## Summary by CodeRabbit

* **Tests**
  * Expanded MIDI CC mapping coverage and hardened behavior checks (channel filtering, unmapped CCs, continuous scaling and midpoint cases, toggle behavior, and sequential updates).
  * Added on-disk persistence tests with backup/restore and save/load round-trip integrity; validated missing/invalid/corrupt config fallbacks.
  * Strengthened assertions for default mappings, duplicate-CC replacement, add/clear/remove semantics, bulk operations, and learn-status formatting.

[Review Change Stack](https://app.coderabbit.ai/change-stack/sudip-mondal-2002/Amplitron/pull/210?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: tests/test_midi_manager.cpp [hunk 1]
```diff
@@ -310,69 +278,27 @@ TEST(midi_json_roundtrip) {
...
-    ASSERT_EQ(midi.mappings()[2].cc_number, 64);
-    ASSERT_EQ(static_cast<int>(midi.mappings()[2].target_type),
-              static_cast<int>(MidiTargetType::EffectBypass));
-    ASSERT_EQ(static_cast<int>(midi.mappings()[2].mode),
-              static_cast<int>(MidiMappingMode::Toggle));
-
-    // CC74 = Wah sweep
-    ASSERT_EQ(midi.mappings()[3].cc_number, 74);
-    ASSERT_EQ(midi.mappings()[3].effect_name, std::string("WahPedal"));
-    ASSERT_EQ(midi.mappings()[3].param_name, std::string("Sweep"));
```
Comment (sudip-mondal-2002) [inline tests/test_midi_manager.cpp:369]: Why so many Assetions are removed?

File: tests/test_midi_manager.cpp [hunk 2]
```diff
@@ -310,69 +278,27 @@ TEST(midi_json_roundtrip) {
...
-    ASSERT_EQ(midi.mappings()[1].cc_number, 74);
-    ASSERT_EQ(midi.mappings()[1].effect_name, std::string("WahPedal"));
-    ASSERT_EQ(midi.mappings()[1].param_name, std::string("Sweep"));
-
-    ASSERT_EQ(midi.mappings()[2].cc_number, 64);
-    ASSERT_EQ(static_cast<int>(midi.mappings()[2].mode),
-              static_cast<int>(MidiMappingMode::Toggle));
-
-    // Test remove
     midi.remove_mapping(1);
```
Comment (sudip-mondal-2002) [inline tests/test_midi_manager.cpp:349]: here also

**Review by sudip-mondal-2002 — CHANGES_REQUESTED** [review]

File: tests/test_midi_manager.cpp (hunk 2 above)
Comment (rashika-developer) [reply tests/test_midi_manager.cpp:349]: Hi @sudip-mondal-2002 

Thank you for the review! I completely understand your point. 

I have fully restored all of the detailed, field-level data validation assertions across both of the affected tests:
1. **`midi_default_mappings`**: Restored full layout verification for all 4 default mappings (CC 7, 11, 64, and 74), including checking their specific `target_type` and `mode` states.
2. **`midi_json_roundtrip`**: Restored the thorough data-fidelity assertions for the `m1`, `m2`, and `m3` mapping blocks.

**Comment by sudip-mondal-2002** [comment]
@rashika-developer 

THE TEST COVERAGE ISN'T EXHAUSTIVE, PLEASE ACHIEVE AT LEAST 90% TEST COVERAGE IN EVERY FILE

image

**Comment by rashika-developer** [comment]
@sudip-mondal-2002 sir, 
    I've tried to increase the test coverage, but due to platform dependent code, I couldn't raise it above the present state. Thank you sir, for your patience and guidance.
image

File: tests/test_midi_manager.cpp [hunk 3]
```diff
@@ -413,3 +420,265 @@ TEST(midi_learn_cancel) {
...
+    }
+
+    MidiManager mgr;
+    mgr.clear_mappings();
+    mgr.load_config();
+    
+    ASSERT_EQ(static_cast<int>(mgr.mappings().size()), 2);
+    ASSERT_EQ(mgr.mappings()[0].cc_number, 7);
+    ASSERT_EQ(static_cast<int>(mgr.mappings()[0].target_type), static_cast<int>(MidiTargetType::EffectParam));
+}
```
[the remark this answers is not in the document]
Comment (rashika-developer) [reply tests/test_midi_manager.cpp:520]: Thank you for reviewing! I checked the implementation in midi_manager_persist.cpp, 
and I believe the current tests are correct. When load_config() encounters a missing 
or invalid file, it calls install_default_mappings() as a fallback, so the expected 
behavior is 2 default mappings, not empty.

The tests verify this graceful degradation works properly. However, if I'm 
misunderstanding how the code works, I'm happy to learn and correct it!

[the remark this answers is not in the document]
Comment (rashika-developer) [reply tests/test_midi_manager.cpp:520]: Thanks for the review! I ran the tests and they actually pass with the original 
assertions. 
It turns out load_config() DOES install 2 default mappings on missing 
file, invalid JSON, and schema errors—the implementation has this fallback built in. 
The tests were correct from the start.

File: tests/test_midi_manager.cpp [hunk 4]
```diff
@@ -360,26 +360,18 @@ TEST(midi_json_roundtrip) {
...
-
-    ASSERT_EQ(midi.mappings()[1].cc_number, 11);
-    ASSERT_EQ(static_cast<int>(midi.mappings()[1].target_type),
-              static_cast<int>(MidiTargetType::InputGain));
-
-    ASSERT_EQ(midi.mappings()[2].cc_number, 64);
-    ASSERT_EQ(static_cast<int>(midi.mappings()[2].target_type),
-              static_cast<int>(MidiTargetType::EffectBypass));
-    ASSERT_EQ(static_cast<int>(midi.mappings()[2].mode),
-              static_cast<int>(MidiMappingMode::Toggle));
```
Comment (sudip-mondal-2002) [inline tests/test_midi_manager.cpp]: Why so many assertions removed?

**Review by sudip-mondal-2002 — CHANGES_REQUESTED** [review]

**Review by sudip-mondal-2002 — APPROVED** [review]

File: tests/test_midi_manager.cpp (hunk 1 above)
Comment (rashika-developer) [reply tests/test_midi_manager.cpp:369]: Hi @sudip-mondal-2002,

That was an oversight during the initial refactoring of the tests. As you saw in the subsequent commits, I fully restored all of those detailed, field-level data validation assertions to ensure the test suite maintains the high level of precision and coverage the project requires.

Thank you for catching that and for the guidance on keeping the assertions robust!
